File: backend/app/schemas/world.py
import uuid
from datetime import datetime

from pydantic import BaseModel, Field


class WorldBookCreate(BaseModel):
    name: str = Field(..., max_length=100)
    description: str | None = Field(None, max_length=500)
    setting: str | None = Field(None, max_length=10000)
    rules: str | None = Field(None, max_length=10000)
    lore: str | None = Field(None, max_length=10000)
    factions: list[str] | None = Field(None, max_length=50)
    tags: list[str] | None = Field(None, max_length=20)
    cover_image: str | None = None
    is_preset: bool = False


class WorldBookUpdate(BaseModel):
    name: str | None = Field(None, max_length=100)
    description: str | None = Field(None, max_length=500)
    setting: str | None = Field(None, max_length=10000)
    rules: str | None = Field(None, max_length=10000)
    lore: str | None = Field(None, max_length=10000)
    factions: list[str] | None = Field(None, max_length=50)
    tags: list[str] | None = Field(None, max_length=20)
    cover_image: str | None = None
    is_preset: bool | None = None


class TagCount(BaseModel):
    name: str
    count: int


class WorldBookSummary(BaseModel):
    id: uuid.UUID
    name: str
    description: str | None
    tags: list[str] | None
    cover_image: str | None
    is_preset: bool
    created_by: uuid.UUID | None = None
    character_count: int = 0
    created_at: datetime

    model_config = {"from_attributes": True}


class WorldBookDetail(BaseModel):
    id: uuid.UUID
    name: str
    description: str | None
    setting: str | None
    rules: str | None
    lore: str | None
    factions: list[str] | None
    tags: list[str] | None
    cover_image: str | None
    is_preset: bool
    created_by: uuid.UUID | None = None
    character_count: int = 0
    created_at: datetime
    updated_at: datetime

    model_config = {"from_attributes": True}

File: frontend/src/types/world.ts
export interface WorldBook {
  id: string
  name: string
  description: string | null
  setting: string | null
  rules: string | null
  lore: string | null
  factions: string[] | null
  tags: string[] | null
  cover_image: string | null
  is_preset: boolean
  created_by: string | null
  character_count: number
  created_at: string
  updated_at: string
}

export interface WorldBookSummary {
  id: string
  name: string
  description: string | null
  tags: string[] | null
  cover_image: string | null
  is_preset: boolean
  created_by: string | null
  character_count: number
  created_at: string
}

export interface TagCount {
  name: string
  count: number
}

export interface WorldBookCreate {
  name: string
  description?: string
  setting?: string
  rules?: string
  lore?: string
  factions?: string[]
  tags?: string[]
  cover_image?: string
  is_preset?: boolean
}

export interface WorldBookUpdate {
  name?: string
  description?: string
  setting?: string
  rules?: string
  lore?: string
  factions?: string[]
  tags?: string[]
  cover_image?: string
  is_preset?: boolean
}

export interface Character {
  id: string
  world_id: string
  name: string
  avatar: string | null
  identity: string | null
  personality: string | null
  background: string | null
  relationships: string | null
  language_style: string | null
  knowledge: string | null
  greeting: string | null
  tags: string[] | null
  created_at: string
  updated_at: string
}

export interface CharacterSummary {
  id: string
  name: string
  avatar: string | null
  identity: string | null
  tags: string[] | null
}

export interface CharacterCreate {
  name: string
  avatar?: string
  identity?: string
  personality?: string
  background?: string
  relationships?: string
  language_style?: string
  knowledge?: string
  greeting?: string
  tags?: string[]
}

export interface CharacterUpdate {
  name?: string
  avatar?: string
  identity?: string
  personality?: string
  background?: string
  relationships?: string
  language_style?: string
  knowledge?: string
  greeting?: string
  tags?: string[]
}
